src/spi.rs

```diff
@@ -221,13 +222,94 @@ impl<REMAP, PINS> Spi<SPI3, REMAP, PINS> {
         REMAP: Remap<Periph = SPI3>,
         PINS: Pins<REMAP, POS>,
     {
-        Spi::<SPI3, _, _>::_spi(spi, pins, mode, freq.into(), clocks, apb)
+        Spi::<SPI3, _, _, u8>::_spi(spi, pins, mode, freq.into(), clocks, apb)
     }
 }
 
 pub type SpiRegisterBlock = crate::pac::spi1::RegisterBlock;
 
-impl<SPI, REMAP, PINS> Spi<SPI, REMAP, PINS>
+pub trait SpiReadWrite<T> {
+    fn read_data_reg(&mut self) -> T;
+    fn write_data_reg(&mut self, data: T);
+    fn write_fast(&mut self, words: &[T]) -> Result<(), Error>;
+}
+
+impl<SPI, REMAP, PINS, FrameSize> SpiReadWrite<FrameSize> for Spi<SPI, REMAP, PINS, FrameSize>
+where
+    SPI: Deref<Target = SpiRegisterBlock>,
+    FrameSize: Copy,
+{
+    fn read_data_reg(&mut self) -> FrameSize {
+        // NOTE(read_volatile) read only 1 byte (the svd2rust API only allows
+        // reading a half-word)
+        return unsafe { ptr::read_volatile(&self.spi.dr as *const _ as *const FrameSize) };
+    }
+
+    fn write_data_reg(&mut self, data: FrameSize) {
+        // NOTE(write_volatile) see note above
+        unsafe { ptr::write_volatile(&self.spi.dr as *const _ as *mut FrameSize, data) }
+    }
+
+    // Implement write as per the "Transmit only procedure" page 712
+    // of RM0008 Rev 20. This is more than twice as fast as the
+    // default Write<> implementation (which reads and drops each
+    // received value)
+    fn write_fast(&mut self, words: &[FrameSize]) -> Result<(), Error> {
```

Comment:
Since we don't have a write_slow or similar, I think we could rename this to just `write` as it is currently, and remove the point in the comment about this being faster than something else.

Reply:
I was thinking to differentiate it from the `embedded_hal::blocking::spi::Write<T>` function `fn write(&mut self, words: &[T])`, which calls `write_fast()`. There is a naming conflict if they're both named `write`. Perhaps `spi_write` or something similar would be more appropriate?

Reply:
Oh right, that's a good point. I think spi_write might be a bit clearer.